Failure to call parse_json on fields of rows

**Describe the bug**

I am trying to call `parse_json` on a value projected from a `ROW`, and encountering a rust compilation error.

```sql
create table t (r ROW(j VARCHAR));
create materialized view v as select parse_json(t.r.j) from t;
```

error:
```
   Compiling project-9f73e38456a3a0b851595784f13ffd4e830c59b9e4911a0c0846387324ab601e v0.1.0 (/working-dir/rust-compilation/projects/project-9f73e38456a3a0b851595784f13ffd4e830c59b9e4911a0c0846387324ab601e)
error[E0308]: mismatched types
   --> projects/project-9f73e38456a3a0b851595784f13ffd4e830c59b9e4911a0c0846387324ab601e/src/main.rs:711:23
    |
711 |             Tup1::new(parse_json_s((*t_32).0.clone().unwrap().0))
    |             --------- ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `Option<Variant>`, found `Variant`
    |             |
    |             arguments to this function are incorrect
    |
    = note: expected enum `std::option::Option<feldera_sqllib::Variant>`
               found enum `feldera_sqllib::Variant`
note: associated function defined here
   --> /database-stream-processor/crates/dbsp/src/utils/tuple/mod.rs:16:1
    |
16  | / crate::declare_tuples! {
17  | |     Tup1<T1>,
18  | |     Tup2<T1, T2>,
19  | |     Tup3<T1, T2, T3>,
...   |
26  | |     Tup10<T1, T2, T3, T4, T5, T6, T7, T8, T9, T10>,
27  | | }
    | |_^
    = note: this error originates in the macro `crate::declare_tuples` (in Nightly builds, run with -Z macro-backtrace for more info)
help: try wrapping the expression in `Some`
    |
711 |             Tup1::new(Some(parse_json_s((*t_32).0.clone().unwrap().0)))
    |                       +++++                                          +

For more information about this error, try `rustc --explain E0308`.
error: could not compile `project-9f73e38456a3a0b851595784f13ffd4e830c59b9e4911a0c0846387324ab601e` (bin "project-9f73e38456a3a0b851595784f13ffd4e830c59b9e4911a0c0846387324ab601e") due to 1 previous error
```

**To Reproduce**
1. Paste above SQL into the live web console.

**Expected behavior**
1. Generated Rust code successfully compiles

**Context (please complete the following information):**
 - Environment: Cloud Sandbox on 1/29/2025
 - Pipeline Configuration/SQL code: see above

**Additional context**
Seems potentially related to struct field nullability, which @mihaibudiu recently worked on. Thanks again for your help adding support for the nested structs/maps!

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/backend/rust/ToRustInnerVisitor.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/backend/rust/ToRustInnerVisitor.java
@@ -128,6 +128,7 @@ import org.dbsp.sqlCompiler.ir.type.derived.DBSPTypeRawTuple;
 import org.dbsp.sqlCompiler.ir.type.derived.DBSPTypeRef;
 import org.dbsp.sqlCompiler.ir.type.derived.DBSPTypeStruct;
 import org.dbsp.sqlCompiler.ir.type.derived.DBSPTypeTuple;
+import org.dbsp.sqlCompiler.ir.type.derived.DBSPTypeTupleBase;
 import org.dbsp.sqlCompiler.ir.type.primitive.DBSPTypeAny;
 import org.dbsp.sqlCompiler.ir.type.primitive.DBSPTypeBaseType;
 import org.dbsp.sqlCompiler.ir.type.primitive.DBSPTypeBinary;
@@ -1601,6 +1602,7 @@ public class ToRustInnerVisitor extends InnerVisitor {
                         bin.opcode == DBSPOpcode.MAP_INDEX ||
                         bin.opcode == DBSPOpcode.VARIANT_INDEX) {
                     if (this.visitingChild == 0)
+                        // We are explicitly taking a reference for these cases
                         avoidRef = true;
                 }
             } else if (parent.is(DBSPBorrowExpression.class)) {
@@ -1609,8 +1611,15 @@ public class ToRustInnerVisitor extends InnerVisitor {
             }
         }
 
+        // Expression of the form x.0
+        // We have four cases, depending on the nullability of x, and of the 0-th field
+        // in the type of x.  Invariant:
+        // we generate code such that
+        // - if the field is further indexed, we return a reference to the field (as_ref())
+        // - if the field is a structure, we return a reference (as_ref()).  We expect
+        //   that the consumer will clone this if necessary.
+        // - if the field is a scalar, we return the field's value (cloned())
         DBSPType sourceType = expression.expression.getType();
-        /*
         boolean fieldTypeIsNullable = false;
         if (sourceType.is(DBSPTypeTupleBase.class)) {
             // This should always be true when we compile SQL programs, but
@@ -1618,37 +1627,8 @@ public class ToRustInnerVisitor extends InnerVisitor {
             DBSPType fieldType = sourceType.to(DBSPTypeTupleBase.class).getFieldType(expression.fieldNo);
             fieldTypeIsNullable = fieldType.mayBeNull;
         }
-         */
-        if (sourceType.mayBeNull) {
-            // Accessing a nullable struct is allowed
-            if (!expression.getType().mayBeNull) {
-                // If the result is not null, we need to unwrap()
-                expression.expression.accept(this);
-                this.builder.append(".as_ref().unwrap().")
-                        .append(expression.fieldNo);
-            } else {
-                expression.expression.accept(this);
-                if (!expression.expression.is(DBSPFieldExpression.class))
-                    this.builder.append(".as_ref()");
-                this.builder.increase().append(".and_then(|x");
-                /*
-                if (this.options.ioOptions.verbosity > 0) {
-                    this.builder.append(": ");
-                    sourceType.withMayBeNull(false).ref().accept(this);
-                }
-                 */
-                this.builder.append("| ");
-                // Indexing a nullable expression with a non-nullable field
-                // if (!fieldTypeIsNullable) this.builder.append("Some(");
-                this.builder.append("x.")
-                        .append(expression.fieldNo);
-                // if (!fieldTypeIsNullable) this.builder.append(")");
-                if (expression.getType().mayBeNull && !expression.getType().hasCopy() && !avoidRef) {
-                    this.builder.append(".as_ref()");
-                }
-                this.builder.append(")").decrease();
-            }
-        } else {
+        if (!sourceType.mayBeNull) {
+            // x is not nullable
             expression.expression.accept(this);
             this.builder.append(".")
                     .append(expression.fieldNo);
@@ -1656,8 +1636,51 @@ public class ToRustInnerVisitor extends InnerVisitor {
                 // The value is in an option.
                 this.builder.append(".as_ref()");
                 if (expression.getType().hasCopy()) {
+                    // This cloned() is used to convert Option<&T> into a T value.
+                    // It is done only for scalar Ts, because all other expressions
+                    // are followed by a real clone() call, which will be also compiled into a cloned().
                     this.builder.append(".cloned()");
                 }
+            }
+        } else {
+            // Accessing a field within a nullable struct is allowed
+            if (!expression.getType().mayBeNull) {
+                // If the result is not null, we need to unwrap().
+                // This should really not happen.
+                assert false;
+
+                this.compiler.reportWarning(expression.getSourcePosition(), "Potential crash",
+                        "Expecting a non-null field from a possibly nullable struct");
+                expression.expression.accept(this);
+                this.builder.append(".unwrap().")
+                        .append(expression.fieldNo);
+            } else {
+                expression.expression.accept(this);
+                if (!expression.expression.is(DBSPFieldExpression.class))
+                    this.builder.append(".as_ref()");
+                this.builder.increase();
+                if (fieldTypeIsNullable) {
+                    this.builder.append(".and_then(|x| x.");
+                } else {
+                    this.builder.append(".map(|x| ");
+                    if (!expression.getType().hasCopy())
+                        this.builder.append("&");
+                    this.builder.append("x.");
+                }
+                /*
+                if (this.options.ioOptions.verbosity > 0) {
+                    this.builder.append(": ");
+                    sourceType.withMayBeNull(false).ref().accept(this);
+                }
+                 */
+                this.builder.append(expression.fieldNo);
+                if (fieldTypeIsNullable &&
+                        expression.getType().mayBeNull &&
+                        !expression.getType().hasCopy() &&
+                        !avoidRef) {
+                    this.builder.append(".as_ref()");
+                }
+                this.builder.append(")").decrease();
             }
         }
         this.pop(expression);

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/outer/LowerCircuitVisitor.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/outer/LowerCircuitVisitor.java
@@ -55,10 +55,10 @@ public class LowerCircuitVisitor extends CircuitCloneVisitor {
      * @param flatmap  Flatmap operation to rewrite. */
     public static DBSPExpression rewriteFlatmap(DBSPFlatmap flatmap) {
         //   move |x: &Tuple2<Array<i32>, Option<i32>>, | -> _ {
-        //     let x0: Vec<i32> = (*x.0).clone();
+        //     let x0: Array<i32> = (*x.0).clone();
         //     let x1: x.1.clone();
-        //     let array = (*x).0.clone();
-        //     array.clone().into_iter().map({
+        //     let array_clone = (*x).0.clone();
+        //     array_clone.into_iter().map({
         //        move |e: i32, | -> Tuple3<Vec<i32>, Option<i32>, i32> {
         //            Tuple3::new(x0.clone(), x1.clone(), e)
         //        }

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/ir/expression/DBSPFieldExpression.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/ir/expression/DBSPFieldExpression.java
@@ -43,6 +43,7 @@ public final class DBSPFieldExpression extends DBSPExpression {
         this.expression = expression;
         this.fieldNo = fieldNo;
         assert fieldNo >= 0: "Negative field index " + fieldNo;
+        assert !expression.getType().mayBeNull || type.mayBeNull;
     }
 
     DBSPFieldExpression(DBSPExpression expression, int fieldNo, DBSPType type) {
@@ -54,11 +55,9 @@ public final class DBSPFieldExpression extends DBSPExpression {
             return type;
         DBSPTypeTupleBase tuple = type.to(DBSPTypeTupleBase.class);
         DBSPType fieldType = tuple.getFieldType(fieldNo);
-        /*
-        TODO: https://github.com/feldera/feldera/issues/3418
+        // A Field access in a nullable struct is nullable
         if (type.mayBeNull)
             fieldType = fieldType.withMayBeNull(true);
-         */
         return fieldType;
     }
 

--- a/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/RegressionTests.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/RegressionTests.java
@@ -274,6 +274,31 @@ public class RegressionTests extends SqlIoTest {
     }
 
     @Test
+    public void issue3814b() {
+        var ccs = this.getCCS("""
+                CREATE TYPE person AS (
+                    id INTEGER,
+                    name TEXT);
+                
+                CREATE TABLE example_table(data person ARRAY);
+                
+                CREATE VIEW V AS SELECT element.id
+                FROM example_table,
+                UNNEST(data) AS element;""");
+        ccs.step("""
+                INSERT INTO example_table (data) VALUES
+                (ARRAY[ROW(1, 'Alice')::person, ROW(2, 'Bob')::person, NULL, ROW(NULL, 'Charlie')::person]),
+                (ARRAY[ROW(3, 'David')::person, ROW(NULL, NULL)::person]);""", """
+                 id | weight
+                -------------
+                 1  | 1
+                 2  | 1
+                    | 3
+                 3  | 1""");
+        this.addRustTestCase(ccs);
+    }
+
+    @Test
     public void issue3095() {
         var ccs = this.getCCS("""
                 CREATE FUNCTION udf(input INT) RETURNS INT;

--- a/sql-to-dbsp-compiler/build.sh
+++ b/sql-to-dbsp-compiler/build.sh
@@ -13,7 +13,7 @@ if true; then
     # This is the standard behavior
     CALCITE_REPO="https://github.com/apache/calcite.git"
     CALCITE_BRANCH="main"
-    CALCITE_NEXT_COMMIT="393ccd1afddf57ab55440ea7ee5b98c9a0092483"
+    CALCITE_NEXT_COMMIT="53216ad33fd0c5ef53086ca32e1871eb21d8bed0"
 else
     # Switch to this script when testing a branch in mihaibudiu's fork that
     # hasn't been merged yet

